File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        Map<String, Integer> completionMap = new HashMap<>();
        for (String s : completion) {
            int prevValue = completionMap.getOrDefault(s, 0);
            
            completionMap.put(s, prevValue + 1);
        }
        
        for (String s : participant) {
            int value = completionMap.getOrDefault(s, 0);
            
            if (value == 0) {
                return s;
            }
            completionMap.put(s, value - 1);
        }
        
        return null;
    }
}

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.js
function solution(participant, completion) {
    const completionMap = new Map();
    
    for (const c of completion) {
        completionMap.set(c, (completionMap.get(c) ?? 0) + 1);
    }
    
    for (const p of participant) {
        if (!completionMap.has(p)) {
            return p;
        }
        
        const count = completionMap.get(p);
        if (count <= 1) {
            completionMap.delete(p);
        } else {
            completionMap.set(p, count - 1);
        }
    }
    
    return null;
}
